add streaming capability to FakeMessagesListChatModel

thehunmonkgroup opened the pull request:
Attempt to add streaming capability to `FakeMessagesListChatModel`.

This is an extension of the work from #10152 and #10116

Not sure I've implemented this correctly, but it is working in my testing.

@hwchase17 happy to make any adjustments if the PR needs improvement.

thehunmonkgroup commented:
@hwchase17 any feedback on this? I'm using this new feature in all my testing and it's working very well.

efriis commented on libs/langchain/langchain/chat_models/fake.py:
```diff
@@ -16,28 +16,77 @@
         **kwargs: Any,
     ) -> ChatResult:
+        response = self._call(messages, stop=stop, run_manager=run_manager, **kwargs)
+        generation = ChatGeneration(message=response)
```
I suspect this may have weird behavior if response is a list of chunks

thehunmonkgroup replied:
I don't think so, I'm using this in one of my tests and it works fine:

```python
    function_responses = [
        [
            AIMessageChunk(
                content="",
                additional_kwargs={"function_call": {"name": "test_function", "arguments": ""}},
            ),
            AIMessageChunk(
                content="",
                additional_kwargs={
                    "function_call": {"arguments": '{\n  "word": "foo",\n  "repeats": 2\n}'}
                },
            ),
        ],
        [
            AIMessageChunk(content="Foo repeated twice is: "),
            AIMessageChunk(content="foo foo"),
        ],
    ]
```

efriis commented on libs/langchain/langchain/chat_models/fake.py:
```diff
@@ -16,28 +16,77 @@
+        run_manager: Optional[CallbackManagerForLLMRun] = None,
+        **kwargs: Any,
+    ) -> Union[BaseMessage, List[BaseMessage]]:
+        """First try to lookup in queries, else return 'foo' or 'bar'."""
```
Update docstring to reflect that it's rotating through

efriis commented on libs/langchain/langchain/chat_models/fake.py:
```diff
@@ -16,28 +16,77 @@
+            self.i += 1
+        else:
+            self.i = 0
+        for c in response:
```
again, not sure this gracefully handles both types for elements of self.responses.

thehunmonkgroup replied:
Unless I'm misunderstanding, I believe it does, this also works in my tests:

```python
    function_responses = [
        AIMessage(
            content="",
            additional_kwargs={
                "function_call": {
                    "name": "test_function",
                    "arguments": '{\n  "word": "foo",\n  "repeats": 2\n}',
                }
            },
        ),
        "Foo repeated twice is: foo foo",
    ]
```

efriis reviewed (commented):
This is cool! It needs to tolerate both types though.

My suggestion would be:

if response is a list:
- concat them together into a message for generate
- stream each chunk for stream

if response is a message:
- return the message for generate
- stream a single big chunk message for stream, like some of the other runnables do

efriis commented:
Also sorry for the delay on this one!

thehunmonkgroup commented:
> This is cool! It needs to tolerate both types though.
> 
> My suggestion would be:
> 
> if response is a list:
> 
> * concat them together into a message for generate
> * stream each chunk for stream
> 
> if response is a message:
> 
> * return the message for generate
> * stream a single big chunk message for stream, like some of the other runnables do

I'm not sure I understand the limitation you're pointing to. The examples I offered in other comments work well in my testing, if that's not what you mean, can you offer a specific example of a case my code doesn't handle?

thehunmonkgroup commented:
@efriis bump -- this was one of the PRs that accidentally got closed.

I handled the merge conflicts due to the relocation of the class into `langchain_community` and re-tested it -- still works flawlessly on my rather complicated test cases.

I believe I've addressed your concerns above, but if not, please do clarify -- would love to finally get this merged!

thehunmonkgroup commented:
@efriis bump again. Fixed the type-checking and linting errors, and ran this through my existing test suite, as far as I can tell all cases are now handled. What else do we need to get this in?

nfcampos commented:
@thehunmonkgroup thanks for this, could you add some tests for the new functionality? Namely calling both invoke and stream on this fake chat model with all possible types of `responses`:
- list of message
- list of str
- list of list of str
- list of list of message

thehunmonkgroup committed:
add streaming capability to FakeMessagesListChatModel

thehunmonkgroup commented:
@nfcampos requested tests added.

Please, for the love of all things holy, land this PR ;) This has been a real exercise in stamina...

nfcampos committed:
Add failing test

nfcampos commented:
@thehunmonkgroup hi, it seems your code changes mean some `responses` types only work for streaming, and some types only work for invoke, we can't have that, all types need to work for all methods. I've added a failing unit test that shows this

thehunmonkgroup commented:
@nfcampos if you really want to do that, then we need to make some decisions about how. I've spent a LOT of time on this PR, and I don't want to do any more work unless I have a clear picture of what is needed to get this committed.

The entire purpose of having the two different types (lists, and lists of lists), is to support multiple responses for both streaming chunks and non-streaming full responses.

In a testing environment, it doesn't really make sense to test non-streaming with lists of lists, and streaming with lists. But if you really want to support all types for both streaming and non-streaming, then we need to decide how we'd transform lists of lists if they were used for non-streaming, and lists if they were used for streaming.

I would propose this:

* For non-streaming requests, if a response is a list of lists, then the inner list would have all elements concatenated into a single response string.
* For streaming requests, if a response is a list, then each list element would be wrapped inside a list.

This would enabled a regular list to be passed for streaming, and it would stream only a single chunk for that response, and for non-streaming, all chunks would be put together into a single response.

Will this approach be sufficient to meet your requirements?

Commit:
x

Commit:
Merge branch 'master' into add-streaming-to-fake-messages-list-chat-model

Commit:
x

Commit:
x

efriis commented:
after playing a bit, this actually has some pretty big outstanding issues related to what happens when a "streaming expected" response gets invoked, or a non-streaming expected response gets streamed.

closing for now. The tests and new types from code review should be kept as they are now if someone reopens
